packages/eslint-plugin/src/rules/prefer-destructuring.ts

```diff
@@ -0,0 +1,239 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import type { JSONSchema4 } from '@typescript-eslint/utils/json-schema';
+import * as tsutils from 'ts-api-utils';
+import type * as ts from 'typescript';
+
+import type {
+  InferMessageIdsTypeFromRule,
+  InferOptionsTypeFromRule,
+} from '../util';
+import { createRule, getParserServices, isTypeAnyType } from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+const baseRule = getESLintCoreRule('prefer-destructuring');
+
+type BaseOptions = InferOptionsTypeFromRule<typeof baseRule>;
+type EnforcementOptions = BaseOptions[1] & {
+  enforceForDeclarationWithTypeAnnotation?: boolean;
+};
+type Options = [BaseOptions[0], EnforcementOptions?];
+
+type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
+
+const destructuringTypeConfig: JSONSchema4 = {
+  type: 'object',
+  properties: {
+    array: {
+      type: 'boolean',
+    },
+    object: {
+      type: 'boolean',
+    },
+  },
+  additionalProperties: false,
+};
+
+const schema: readonly JSONSchema4[] = [
+  {
+    oneOf: [
+      {
+        type: 'object',
+        properties: {
+          VariableDeclarator: destructuringTypeConfig,
+          AssignmentExpression: destructuringTypeConfig,
+        },
+        additionalProperties: false,
+      },
+      destructuringTypeConfig,
+    ],
+  },
+  {
+    type: 'object',
+    properties: {
+      enforceForRenamedProperties: {
+        type: 'boolean',
+      },
+      enforceForDeclarationWithTypeAnnotation: {
+        type: 'boolean',
+      },
+    },
+  },
+];
+
+export default createRule<Options, MessageIds>({
+  name: 'prefer-destructuring',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Require destructuring from arrays and/or objects',
+      extendsBaseRule: true,
+      requiresTypeChecking: true,
+    },
+    schema,
+    fixable: baseRule.meta.fixable,
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [
+    {
+      VariableDeclarator: {
+        array: true,
+        object: true,
+      },
+      AssignmentExpression: {
+        array: true,
+        object: true,
+      },
+    },
+    {},
+  ],
+  create(context, [enabledTypes, options1 = {}]) {
```

Comment:
[Naming] IMO there's no need to call this `options1`; it's clear from the code context that it's the rule's `options`.

```suggestion
  create(context, [enabledTypes, options = {}]) {
```